syntax = "proto3";

package mpay;

service Mpay {
  rpc GetInfo (GetInfoRequest) returns (GetInfoResponse) {}

  rpc GetRoutes (GetRoutesRequest) returns (GetRoutesResponse) {}
  rpc ListPayments (ListPaymentsRequest) returns (ListPaymentsResponse) {}

  rpc Pay (PayRequest) returns (PayResponse) {}

  rpc ResetPathMemory (ResetPathMemoryRequest) returns (ResetPathMemoryResponse) {}

  // Workaround to expose the paystatus command via gRPC, since CLN doesn't
  rpc PayStatus (PayStatusRequest) returns (PayStatusResponse) {}
}

message GetInfoRequest {}
message GetInfoResponse {
  string version = 1;
}

message GetRoutesRequest {
  optional string destination = 1;
  optional float min_success = 2;
  optional float min_success_ema = 3;
}

message GetRoutesResponse {
  message Routes {
    message Route {
      repeated string route = 1;
      double success_rate = 2;
      double success_rate_ema = 3;
    }

    repeated Route routes = 2;
  }

  map<string, Routes> routes = 1;
}

message ListPaymentsRequest {
  oneof identifier {
    string bolt11 = 1;
    string payment_hash = 2;
    PaginationParams pagination = 3;
  }
}

message PaginationParams {
  uint64 offset = 1;
  uint32 limit = 2;
}

message ListPaymentsResponse {
  message Payment {
    message Attempt {
      message Hop {
        uint64 id = 1;
        string node = 2;
        string channel = 3;
        uint64 direction = 4;
        bool ok = 5;
      }

      uint64 id = 1;
      bool ok = 2;
      uint64 time = 3;
      repeated Hop hops = 4;
      uint64 created_at = 5;
    }

    uint64 id = 1;
    string destination = 2;
    string payment_hash = 3;
    uint64 amount = 4;
    bool ok = 5;
    repeated Attempt attempts = 6;
    uint64 created_at = 7;
  }

  repeated Payment payments = 1;
}

message PayRequest {
  string bolt11 = 1;
  optional uint64 max_fee_msat = 2;
  optional uint64 exempt_fee_msat = 3;
  optional uint64 timeout = 4;
  optional uint64 max_delay = 5;
}

message PayResponse {
  string payment_hash = 1;
  string payment_preimage = 2;
  uint64 fee_msat = 3;
  uint64 time = 4;
  string destination = 5;
  uint64 amount_msat = 6;
  uint64 amount_sent_msat = 7;
  uint32 parts = 8;
  string status = 9;
  uint64 created_at = 10;
}

message ResetPathMemoryRequest {
  optional bool exclude_permanent_memory = 1;
  optional bool exclude_temporary_memory = 2;
}

message ResetPathMemoryResponse {
  uint64 payments = 1;
  uint64 attempts = 2;
  uint64 hops = 3;
}

message PayStatusRequest {
  optional string invoice = 1;
}

message PayStatusResponse {
  message PayStatus {
    message Attempt {
      enum AttemptState {
        ATTEMPT_PENDING = 0;
        ATTEMPT_COMPLETED = 1;
      }

      message Success {
        uint64 id = 1;
        string payment_preimage = 2;
      }

      message Failure {
        message Data {
          uint64 id = 1;
          string raw_message = 2;
          uint64 fail_code = 3;
          string fail_codename = 4;
          uint64 erring_index = 5;
          string erring_node = 6;
        }

        string message = 1;
        uint64 code = 2;
        optional Data data = 3;
      }

      string strategy = 1;
      uint64 start_time = 2;
      uint64 age_in_seconds = 3;
      optional uint64 end_time = 4;
      AttemptState state = 5;
      optional Success success = 6;
      optional Failure failure = 7;
    }

    optional string bolt11 = 1;
    optional string bolt12 = 5;

    uint64 amount_msat = 2;
    string destination = 3;
    repeated Attempt attempts = 4;
  }

  repeated PayStatus status = 1;
}
